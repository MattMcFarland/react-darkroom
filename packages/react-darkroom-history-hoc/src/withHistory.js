    import { withStateHandlers } from 'recompose'

    export default (initialState) => withStateHandlers({ thread: [initialState], index: 0, ...initialState }, {
      pushHistory: (state, props) => newState => updateThread(state, newState),
      go: (state, props) => n => moveIndex(state, n),
      forward: (state, props) => () => moveIndex(state, 1),
      backward: (state, props) => () => moveIndex(state, -1),
      undo: (state, props) => () => moveIndex(state, -1),
      redo: (state, props) => () => moveIndex(state, 1)
    })

    export function updateThread (state, newState) {
      const updateThread = [
        ...state.thread.slice(0, state.index + 1),
        newState
      ];
      const updateIndex = updateThread.length - 1
      return {
        index: updateIndex,
        thread: updateThread,
        ...updateThread[updateIndex]
      }
    }

    export function moveIndex (state, moveBy) {
      const targetIndex = clamp(state.index + moveBy, 0, state.thread.length - 1)
      return {
        index: targetIndex,
        thread: state.thread,
        ...state.thread[targetIndex]
      }
    }

    export function clamp (num, min, max) {
      return num <= min ? min : num >= max ? max : num;
    }

