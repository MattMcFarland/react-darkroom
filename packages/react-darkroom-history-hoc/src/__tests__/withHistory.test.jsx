import React from 'react'
import { shallow, mount, render } from 'enzyme';
import withHistory, { moveIndex, updateThread, clamp } from '../withHistory';

const { describe, test, expect } = global;

const HistoryTest = ({
  counter,
  go,
  pushHistory,  
  undo,
  backward,
  redo,
  forward,
  index,
  thread,
  goByAmount
}) => (
  <div>
    <h1>HistoryTest</h1>
    <p><span>Count:</span><num>{counter}</num></p>
    <button increment onClick={() => pushHistory({counter: counter + 1})}>increment</button>
    <button decrement onClick={() => pushHistory({counter: counter - 1})}>decrement</button>
    <button undo onClick={undo}>Undo</button>
    <button redo onClick={redo}>Redo</button>
    <button backward onClick={backward}>Back</button>
    <button forward onClick={forward}>Forward</button>
    <button go onClick={() => go(goByAmount)}></button>
  </div>
)
const initialState = { counter: 1 };
const TestHistory = withHistory(initialState)(HistoryTest);

const runSimulationOnAttributes = props => (...attributes) => {
  const simulatedWrapper = attributes.reduce((wrapper, attribute) => {
    wrapper.find(`button[${attribute}]`).simulate('click')
    return wrapper 
  }, shallow(<TestHistory {...props}/>))
  // creating snapshots to make it easier to see results, might remove later.
  expect(simulatedWrapper.state()).toMatchSnapshot();
  return simulatedWrapper;
}

describe('withHistory higher order component', () => {
  describe('thread', () => {
    test('thread starts with initalState', () => {
      const wrapper = shallow(<TestHistory/>);
      expect(wrapper.state().thread).toEqual([{ counter: 1 }]);
    })
    test('root state contains data stored in the thread by index', () => {
      const state = runSimulationOnAttributes()('increment', 'decrement', 'undo').state();
      expect(state.counter).toEqual(state.thread[state.index].counter);
    })
  })

  describe('pushHistory()', () => {
    test('pushHistory adds actions called to thread', () => {
      // increment, decrement use pushHistory in this test.
      const state = runSimulationOnAttributes()('increment', 'decrement', 'increment').state();
      expect(state.thread.length).toEqual(4);
    })  
  })

  describe('undo/backward', () => {
    test('subtracts the index, but not the length', () => {
      const state = runSimulationOnAttributes()('increment', 'decrement', 'increment', 'undo', 'backward').state();
      expect(state.thread.length).toEqual(4);
      expect(state.index).toEqual(1)
    })
    test('prevents index from going out of range', () => {
      const state = runSimulationOnAttributes()('undo', 'backward').state();
      expect(state.thread.length).toEqual(1);
      expect(state.index).toEqual(0)      
    })
    test('whenever a new action follows an undo, the thread is cut to the index of the new action', () => {
      const state = runSimulationOnAttributes()('increment', 'increment', 'increment', 'undo', 'undo', 'decrement').state();
      expect(state.index).toEqual(state.thread.length - 1);
    })
  })

  describe('redo/forward', () => {
    test('raises the index when it is less than the length', () => {
      const state = runSimulationOnAttributes()('increment', 'decrement', 'undo', 'redo', 'backward', 'forward').state();
      expect(state.thread.length).toEqual(3);
      expect(state.index).toEqual(2)
    })
    test('prevents index from going out of range', () => {
      const state = runSimulationOnAttributes()('redo', 'forward').state();
      expect(state.thread.length).toEqual(1);
      expect(state.index).toEqual(0)      
    })    
  })
  describe('go() relatively adds the index to change the state', () => {
    test('up', () => {      
      const state = runSimulationOnAttributes({goByAmount: 2})
        (
          'increment', 
          'increment', 
          'increment', 
          'increment',
          'undo',
          'undo',
          'go',
        ).state();      
        expect(state.thread.length).toEqual(5);
        expect(state.index).toEqual(4)           
    })
    test('down', () => {
      const state = runSimulationOnAttributes({goByAmount: -2})
      (
        'increment', 
        'increment', 
        'increment', 
        'increment',
        'go',
      ).state();     
      expect(state.thread.length).toEqual(5);
      expect(state.index).toEqual(2)    
    })

  })
});
