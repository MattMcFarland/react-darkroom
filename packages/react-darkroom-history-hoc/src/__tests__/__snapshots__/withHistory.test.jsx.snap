// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withHistory higher order component go() relatively adds the index to change the state down 1`] = `
Object {
  "counter": 3,
  "index": 2,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 3,
    },
    Object {
      "counter": 4,
    },
    Object {
      "counter": 5,
    },
  ],
}
`;

exports[`withHistory higher order component go() relatively adds the index to change the state up 1`] = `
Object {
  "counter": 5,
  "index": 4,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 3,
    },
    Object {
      "counter": 4,
    },
    Object {
      "counter": 5,
    },
  ],
}
`;

exports[`withHistory higher order component pushHistory() pushHistory adds actions called to thread 1`] = `
Object {
  "counter": 2,
  "index": 3,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
  ],
}
`;

exports[`withHistory higher order component redo/forward prevents index from going out of range 1`] = `
Object {
  "counter": 1,
  "index": 0,
  "thread": Array [
    Object {
      "counter": 1,
    },
  ],
}
`;

exports[`withHistory higher order component redo/forward raises the index when it is less than the length 1`] = `
Object {
  "counter": 1,
  "index": 2,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 1,
    },
  ],
}
`;

exports[`withHistory higher order component thread root state contains data stored in the thread by index 1`] = `
Object {
  "counter": 2,
  "index": 1,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 1,
    },
  ],
}
`;

exports[`withHistory higher order component undo/backward prevents index from going out of range 1`] = `
Object {
  "counter": 1,
  "index": 0,
  "thread": Array [
    Object {
      "counter": 1,
    },
  ],
}
`;

exports[`withHistory higher order component undo/backward subtracts the index, but not the length 1`] = `
Object {
  "counter": 2,
  "index": 1,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
  ],
}
`;

exports[`withHistory higher order component undo/backward whenever a new action follows an undo, the thread is cut to the index of the new action 1`] = `
Object {
  "counter": 1,
  "index": 2,
  "thread": Array [
    Object {
      "counter": 1,
    },
    Object {
      "counter": 2,
    },
    Object {
      "counter": 1,
    },
  ],
}
`;
